main.rb

```diff
@@ -0,0 +1,44 @@
+# def bubble_sort(arr)
+#   number_of_loops = arr.length - 1
+#   swap = true
+#   while swap
+#     swap = false
+#     number_of_loops.times do |x|
+#       arr[x], arr[x + 1] = arr[x + 1], arr[x] if arr[x] > arr[x + 1]
+#       swap = true
+#     end
+#     number_of_loops -= 1
+#   end
+#   arr
+# end
+# puts bubble_sort([3, 2, 1, 85, 1, 58, 6, 4])
+
+# Another way to solve it:
+def bubble_sort(arr)
+  number_ofloop = arr.length - 1 # loop inside the index
+  (0..number_ofloop).each do |_i| # the first outside loop
+    (0...number_ofloop).each do |j| # the inside loop
+      arr[j], arr[j + 1] = arr[j + 1], arr[j] if arr[j] > arr[j + 1] # replacement of the elements
+    end
+  end
+  arr
+end
+
+puts bubble_sort([4, 3, 78, 2, 0, 2])
+
+# The second:
+
+def bubble_sort_by(arr)
+  number_ofloop = arr.length - 1 # loop inside the index not the length
+  (0..number_ofloop).each do |_i|
+    (0...number_ofloop).each do |j|
+      arr[j], arr[j + 1] = arr[j + 1], arr[j] if arr[j].length > arr[j + 1].length
+    end
+  end
+  arr
+end
```

Comment:
- [ ] #bubble_sort_by isn't working, if you try `bubble_sort_by(%w[hi hello hey]) { |left, right| left.length + right.length }` it wouldn't give the right result (`[hello hey hi]`) (Hint: you're not using `yield`)